### DISCUSSION TITLE
Incorrect exact mass calculated by RdKit

### DISCUSSION THREAD

**Discussion by asmirn1** [rdkit/rdkit#4201] [Q&A]
Hi,

I use the following code to calculate the exact mass:

```python
structure_data = PandasTools.LoadSDF(path_to_sdf, molColName='Molecule')
selected_data = structure_data[structure_data['DATABASE_ID'] == 'HMDB0006116']
for _, row in selected_data.iterrows():
    molecule = structure_row['Molecule']
    molecular_mass = Descriptors.ExactMolWt(molecule)
    print(molecular_mass)
```

The calculated exact molecular weight is 187.998. However, the correct molecular weight (returned by PubChem) is 195.17. It looks like RDKit doesn't count hydrogens?

The compound structure from PubChem:
![3-Hydroxyhippuric_acid_300](https://user-images.githubusercontent.com/13631120/120541605-15ad2680-c3b8-11eb-8de2-732f78e387dc.png)

The compound structure plotted by RDKit:
![HMDB0006116](https://user-images.githubusercontent.com/13631120/120541520-fd3d0c00-c3b7-11eb-9e57-276ad2e3ca76.png)

I have also tried using `Chem.AddHs` as follows:
```python
molecule = Chem.AddHs(structure_row['Molecule'])
```
But the exact mass is still incorrect (187.998) and the structure plot looks strange:
![HMDB0006116](https://user-images.githubusercontent.com/13631120/120541966-89e7ca00-c3b8-11eb-8882-3fb97747f2a8.png)

Please, help to find the correct exact weight.

Thank you

**Comment by DavidACosgrove** [comment]
Hi,
The dots in the RDKit drawing denote radicals. I expect there is something
wrong with your input structure.
Regards,
Dave


On Wed, 2 Jun 2021 at 20:39, Aleksandr Smirnov ***@***.***>
wrote:

> Hi,
>
> I use the following code to calculate the exact mass:
>
> structure_data = PandasTools.LoadSDF(path_to_sdf, molColName='Molecule')selected_data = structure_data[structure_data['DATABASE_ID'] == 'HMDB0006116']for _, row in selected_data.iterrows():
>     molecule = structure_row['Molecule']
>     molecular_mass = Descriptors.ExactMolWt(molecule)
>     print(molecular_mass)
>
> The calculated exact molecular weight is 187.998. However, the correct
> molecular weight (returned by PubChem) is 195.17. It looks like RDKit
> doesn't count hydrogens?
>
> The compound structure from PubChem:
> [image: 3-Hydroxyhippuric_acid_300]
> <https://user-images.githubusercontent.com/13631120/120541605-15ad2680-c3b8-11eb-8de2-732f78e387dc.png>
>
> The compound structure plotted by RDKit:
> [image: HMDB0006116]
> <https://user-images.githubusercontent.com/13631120/120541520-fd3d0c00-c3b7-11eb-9e57-276ad2e3ca76.png>
>
> I have also tried using Chem.AddHs as follows:
>
> molecule = Chem.AddHs(structure_row['Molecule'])
>
> But the exact mass is still incorrect (187.998) and the structure plot
> looks strange:
> [image: HMDB0006116]
> <https://user-images.githubusercontent.com/13631120/120541966-89e7ca00-c3b8-11eb-8882-3fb97747f2a8.png>
>
> Please, help to find the correct exact weight.
>
> Thank you
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rdkit/rdkit/discussions/4201>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FQQ2XNZZUFK57LCRRTTQ2CIHANCNFSM457LDNHA>
> .
>
-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

**Reply by bp-kelley** [reply]
The radicals look like the issue to me, starting from the smiles string:

```
>>> from rdkit.Chem import Descriptors
>>> from rdkit.Chem import AllChem
>>> m = AllChem.MolFromSmiles("O=C(O)CNC(=O)c1cccc(O)c1")
>>> Descriptors.MolWt(m)
195.174
```

**Comment by asmirn1** [comment]
@DavidACosgrove @bp-kelley 

I read data from SDF files. So fixing those SDF files is the only option here? Or there are some RDKit functions that I could use to fix the radicals?

Thank you

**Reply by greglandrum** [reply]
@asmirn1 : where did the SDF files come from?
If I download the SDF from the Pubchem record: https://pubchem.ncbi.nlm.nih.gov/compound/3-Hydroxyhippuric-acid
I get a molecule with no radicals and the correct ExactMolWt.

**Reply by asmirn1** [reply]
@greglandrum The SDF files come from a different lab. As far as I know, they compiled them from different sources. I guess we'll try to update those SDF files instead.
